sdk/eventhub/azure-eventhub/azure/eventhub/_buffered_producer/_buffered_producer.py

```diff
@@ -0,0 +1,221 @@
+        # Put single event or EventDataBatch into the queue.
+        # This method would raise OperationTimeout if the queue does not have enough space for the input and
+        # flush cannot finish in timeout.
+        with self._not_full:
+            try:
+                new_events_len = len(events)
+            except TypeError:
+                new_events_len = 1
+            timeout_time = time.time() + timeout if timeout else None
+
+            if self._max_buffer_len - self._cur_buffered_len < new_events_len:
+                _LOGGER.info(
+                    "Partition {} does not have enough room for coming {} events."
+                    "Flush first.".format(
+                        self.partition_id, new_events_len
+                    )
+                )
+                # flush the buffer
+                self.flush(timeout=timeout)
+
+            if timeout_time and time.time() > timeout_time:
+                raise OperationTimeoutError("Failed to enqueue events into buffer due to timeout.")
+
+            try:
+                # add single event into current batch
+                self._cur_batch.add(events)
+            except AttributeError:
+                # if the input events is a EventDataBatch, put the whole into the buffer
+                self._buffered_queue.put(events)
+                # create a new batch for incoming events
+                self._cur_batch = EventDataBatch(self._max_message_size_on_link)
+                # put the new batch into the buffer
+                self._buffered_queue.put(self._cur_batch)
+            except ValueError:
+                # add single event exceeds the cur batch size, create new batch
+                self._cur_batch = EventDataBatch(self._max_message_size_on_link)
+                self._buffered_queue.put(self._cur_batch)
+                self._cur_batch.add(events)
+            self._cur_buffered_len += new_events_len
+            # notify the max_wait_time worker
+            self._not_empty.notify()
+
+    def enhanced_callback_decorator(self, callback):
+        def wrapper_callback(*args, **kwargs):
+            try:
+                callback(*args, **kwargs)
+            except Exception as exc:
+                _LOGGER.warning("callback exception", callback.__name__, exc, self.partition_id)
+
+        return wrapper_callback
+
+    def sent_callback(self, events, future: Future):
+        try:
+            future.result()
+            self._on_success(events, self.partition_id)
+        except Exception as exc:
+            self._on_error(events, exc, self.partition_id)
+        finally:
+            self._cur_buffered_len -= len(events)
+            self._max_concurrent_sends_semaphore.release()
+
+    def flush(self, timeout=None, raise_error=True):
+        # try flushing all the buffered batch within given time
+        _LOGGER.info("Partition: {} started flushing.".format(self.partition_id))
+        with self._not_empty:
```

Comment:
Can I use this Condition of the max_wait_time_worker is already holding/waiting on it?

Reply:
`max_wait_time_worker ` will 
1. acquire `self._not_empty` lock/condition
2. calling `self._not_empty.wait()` will release the lock temporarily, waiting to be notified
3. it will call flush again which within flush it acquires the lock

so in stage 1/3, flush could no use, but in stage2 when `wait`, flush could use.

also underneath this condition is a re-entrant lock, so `max_wait_time_worker` acquire the lock and call flush will work.